lib/cn/node_split.c

```diff
@@ -562,23 +569,40 @@ cn_split(struct cn_compaction_work *w)
     while (inflight > 0)
         usleep(100 * 1000);
 
-    for (i = 0; !err && i < w->cw_kvset_cnt; i++) {
+    for (int64_t i = w->cw_kvset_cnt - 1; !err && i >= 0; i--) {
```

Comment:
Hmmm...  I think it would be more clear and obvious to cast cw_kvset_cnt to an int up front, then I don't have to go through an integer promotion and sign extension rules exercise to verify the correctness :-)

for (int i = (int)w->cw_kvset_cnt - 1; !err && i >= 0; i--) {

Reply:
Just wanted to make sure we handle the case where we got > 2B kvsets in a node :-)
I will change it to what you had suggested!

Reply:
I don't like the fact that you are casting to a signed int from an unsigned int very much. Can the loop logic could be changed to operate with a uint, which w->cw_kvset_cnt is?

Reply:
Think I can do the below and retain 'i' as uint:

```for (i = w->cw_kvset_cnt; !err && i-- > 0;) { }```

Reply:
I was mainly referring to Greg's comment. Though I do not really like promoting integers unless there is a reason. From what I can tell you are only promoting because of the bounds of the for loop unless there is something I am missing.

If you can retain the type of `i`, even better.

Reply:
I prefer to avoid casting if at all possible.  IMO, when you see a cast it should be a red flag that says "Pay Attention, something unsavory is being done here".  Ergo, if one uses casts willy-nilly then it diminishes the value of a cast to stand out and alert the reader to pay attention.

Love the new solution Nabeel, and thanks to Tristan for pushing us there :-)

Reply:
I think it would be a good idea to turn on `-Wconversion` at some point.